File: e2e_testing_bdd/Pages/ResultsPage.cs
using Dynamitey;
using Microsoft.Playwright;
using NUnit.Framework;

namespace e2e_testing_bdd.Pages;

public class ResultsPage
{
    private readonly IPage _page;
    public ResultsPage(IPage page) => _page = page;
    
    private ILocator ResultPageSuccess => _page.Locator("h1:text(\"Look who we found \ud83d\udd25\")");
    private ILocator ResultPageFail => _page.Locator("h2:text(\"Ops, something went wrong \ud83d\ude22\")");
    private ILocator ReturnButton => _page.Locator("button:text(\"Return\")");
    private ILocator ImageElement => _page.Locator("//img[contains(@class,\"avatar\")]");
    private ILocator NameElement => _page.Locator("//b[contains(text(),'Name')]//parent::p[text()]");
    private ILocator NumberOfRepositories => _page.Locator("//b[contains(text(),\"Repositories\")]//parent::p[text()]");
    private ILocator RepositoriesTitles => _page.Locator("//li[contains(@class,\"repoItem\")]/strong");
    private ILocator RepositoriesDescription => _page.Locator("//li[contains(@class,\"repoItem\")]/p");

    public async Task AssertSearchStatus(string status)
    {
        bool isVis;
        await ReturnButton.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Visible });
        switch (status.ToUpper())
        {
            case "SUCCESS":
                isVis = await ResultPageSuccess.IsVisibleAsync();
                Assert.IsTrue(isVis, "Success search was expected.");
                break;
            case "FAIL":
                isVis = await ResultPageFail.IsVisibleAsync();
                Assert.IsTrue(isVis, "Failed search was expected.");
                break;
            default:
                Assert.Fail("Status:" +status+" is not expected.");
                break;
        }
    }

    public async Task AssertElements(Table table)
    {
        foreach (var row in table.Rows)
        {
            var elementName = row["elements"];
            await ReturnButton.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Visible });
            ILocator? elementLocator = null;
            switch (elementName.ToUpper())
            {
                case "IMAGE":
                    elementLocator = ImageElement;
                    break;
                case "NAME":
                    elementLocator = NameElement;
                    break;
                case "NUMBER OF REPOSITORIES":
                    elementLocator = NumberOfRepositories;
                    break;
                case "REPOSITORY TITLE":
                    elementLocator = RepositoriesTitles.First;
                    break;
                case "REPOSITORY DESCRIPTION":
                    elementLocator = RepositoriesDescription.First;
                    break;
                default:
                    Assert.Fail("Element type no expected");
                    break;
            }

            var isVisible = await elementLocator?.IsVisibleAsync()!;
            if (!isVisible)
            {
                Assert.Fail("The element " + elementName + " is not visible");
            }
        }
    }

    public async Task AssertNumberOfRepositories()
    {
        var numberOfRepositories = ((await NumberOfRepositories.TextContentAsync())!).Split(" ")[1];
        var numberOfTitles = await RepositoriesTitles.CountAsync();

        if (numberOfTitles != int.Parse(numberOfRepositories!))
        {
            Console.WriteLine("Number showed: " + numberOfRepositories);
            Console.WriteLine("Number counted: " + numberOfTitles);
            Assert.Fail("The number of repositories shown is different from the ones counted");
        }

    }
}

File: e2e_testing_js/Pages/ResultsPage.js
const { expect } = require('@playwright/test');

exports.ResultsPage=class ResultsPage {
    constructor(page) {
        this.page = page;
        this.resultPageSuccess = page.locator('h1:text("Look who we found 🔥")');
        this.resultPageFail = page.locator('h2:text("Ops, something went wrong 😢")');
        this.returnButton = page.locator('button:text("Return")');
        this.imageElement = page.locator('//img[contains(@class,"avatar")]');
        this.nameElement = page.locator("//b[contains(text(),'Name')]//parent::p[text()]");
        this.numberOfRepositories = page.locator('//b[contains(text(),"Repositories")]//parent::p[text()]');
        this.repositoriesTitles = page.locator('//li[contains(@class,"repoItem")]/strong');
        this.repositoriesDescription = page.locator('//li[contains(@class,"repoItem")]/p');
    }

    async assertSearchStatus(status) {
        await this.returnButton.waitFor({ state: 'visible' });
        let isVis;
        switch (status.toUpperCase()) {
            case 'SUCCESS':
                isVis = await this.resultPageSuccess.isVisible();
                expect(isVis).toBeTruthy();
                break;
            case 'FAIL':
                isVis = await this.resultPageFail.isVisible();
                expect(isVis).toBeTruthy();
                break;
            default:
                throw new Error(`Status: ${status} is not expected.`);
        }
    }

    async assertElementVisible(elementName) {
        await this.returnButton.waitFor({ state: 'visible' });
        let elementLocator;
        switch (elementName.toUpperCase()) {
            case 'IMAGE':
                elementLocator = this.imageElement;
                break;
            case 'NAME':
                elementLocator = this.nameElement;
                break;
            case 'NUMBER OF REPOSITORIES':
                elementLocator = this.numberOfRepositories;
                break;
            case 'REPOSITORY TITLE':
                elementLocator = this.repositoriesTitles.first();
                break;
            case 'REPOSITORY DESCRIPTION':
                elementLocator = this.repositoriesDescription.first();
                break;
            default:
                throw new Error(`Element type ${elementName} not expected.`);
        }

        const isVisible = await elementLocator.isVisible();
        if (!isVisible) {
            throw new Error(`The element ${elementName} is not visible.`);
        }
        
        expect(isVisible).toBeTruthy();
    }

    
    async assertNumberOfRepositories() {
        const numberOfRepositories = (await this.numberOfRepositories.textContent()).split(' ')[1];
        const numberOfTitles = await this.repositoriesTitles.count();

        if (numberOfTitles !== parseInt(numberOfRepositories)) {
            console.log(`Number showed: ${numberOfRepositories}`);
            console.log(`Number counted: ${numberOfTitles}`);
            throw new Error('The number of repositories shown is different from the ones counted');
        }
    }
}

